File: leetcode/114-flatten-binary-tree-to-linked-list.cpp
/*
Given a binary tree, flatten it to a linked list in-place.

For example,
Given

         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:
   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6

*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if (root == nullptr) {
            return;
        }
        stack<TreeNode*> stk;
        while (root->left != nullptr) {
            stk.push(root->right);
            root->right = root->left;
            auto tmp = root;
            root = root->left;
            tmp->left = nullptr;
        }
        stk.push(root->right);
        while (!stk.empty()) {
            auto node = stk.top();
            stk.pop();
            root->right = node;
            if (node == nullptr) {
                continue;
            }
            while (node->left != nullptr) {
                stk.push(node->right);
                node->right = node->left;
                auto tmp = node;
                node = node->left;
                tmp->left = nullptr;
            }
            stk.push(node->right);
            root = node;
        }
    }
};

File: leetcode/114-flatten-binary-tree-to-linked-list.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
def flatten(root)
  _flattern(root)[0]
end


def _flattern(root)
  if root.nil?
    [nil, nil]
  elsif root.left.nil? and root.right.nil?
    root.left = nil
    root.right = nil
    [root, root]
  else
    left_start, left_end = _flattern root.left
    right_start, right_end = _flattern root.right
    root.left = nil
    if left_start.nil?
      root.right = right_start
      [root, right_end]
    else
      root.right = left_start
      left_end.right = right_start
      [root, right_end.nil? ? left_end : right_end]
    end
  end
end
